lib/galaxy/tools/toolbox/base.py

```diff
@@ -828,16 +828,19 @@ def quick_load(tool_file, async_load=True):
                 return None
 
         tool_loaded = False
-        for name in os.listdir(directory):
-            if name.startswith('.' or '_'):
-                # Very unlikely that we want to load tools from a hidden or private folder
-                continue
-            child_path = os.path.join(directory, name)
-            if os.path.isdir(child_path) and recursive:
-                self.__watch_directory(child_path, elems, integrated_elems, load_panel_dict, recursive)
-            elif self._looks_like_a_tool(child_path):
-                quick_load(child_path, async_load=False)
-                tool_loaded = True
+        try:
+            for name in os.listdir(directory):
+                if name.startswith(('.', '_')):
+                    # Very unlikely that we want to load tools from a hidden or private folder
+                    continue
+                child_path = os.path.join(directory, name)
+                if os.path.isdir(child_path) and recursive:
+                    self.__watch_directory(child_path, elems, integrated_elems, load_panel_dict, recursive)
+                elif self._looks_like_a_tool(child_path):
+                    quick_load(child_path, async_load=False)
+                    tool_loaded = True
+        except Exception:
```

Comment:
I'm wondering... This Exception will catch *everything* in this block, including the recursive call. It fails silently - so we'll never know if something bad happens elsewhere. If `directory` doesn't exist, it'll fail silently - which, I think, is the intended behavior. However, if something happens here `self._tool_watcher.watch_directory(directory, quick_load)` on a recursive call, we'll never know - it'll be caught by `Exception`. In other words:
```python
def foo(n):
    if n == -1: 
        return 
    try:
        foo(n-1)
    except Exception:
        print('do nothing')
    bar(n)
def bar(n):
    1/n
foo(1)
```
The above will never catch division by zero in `bar`.

If the goal is to not die on missing directory, maybe we could use something like this on line 831?
```if not os.path.exists(directory): 
        log.exception(...)
        return
``` 
(I assume that's the desired outcome?)

Reply:
The `log.exception()` call will provide the traceback:

```
galaxy.tools.toolbox.base ERROR 2019-12-12 17:01:31,348 [p:17931,w:0,m:0] [MainThread] Failed to read tool directory /path/no/exist/tools.
Traceback (most recent call last):
  File "lib/galaxy/tools/toolbox/base.py", line 832, in __watch_directory
    for name in os.listdir(directory):
OSError: [Errno 2] No such file or directory: '/path/no/exist/tools'
```

The goal is to prevent any error here from causing the rest of the processes to continue.

Reply:
Oh, OK, then that's fine.

Reply:
Maybe instead of the try/except, we can use:
```python
        if not os.path.isdir(directory):
            return
```
before the for loop.